---
title: An R Markdown document converted from "14_GRSCalc.ipynb"
output: html_document
---

# Polygenic score calculation in UKB

```{r}
library(readr)
library(dplyr, warn.conflicts = FALSE)
library(tidyr, warn.conflicts = FALSE)
library(purrr)
library(rbgen)
```

```{r}
for(file_name in list.files("../data/PRSFiles")){
    trait <- gsub("\\.gz", "", file_name)
    full_name <- paste("../data/PRSFiles", file_name, sep = "/")
    summdat <- read_tsv(full_name, comment = "#", show_col_types = FALSE)
    chromosomes <- unique(summdat$chr_name)
    grs_chrom <- list()
    for(chrom in chromosomes){
        chrom_dat <- filter(summdat, chr_name == chrom)
        bgen_filename <- paste0("/ludc/Raw_Data_Archive/UKBB/imp/ukb_imp_chr", chrom, "_v3.bgen")
        bgen_index <- paste0("/ludc/Home/daniel_c/projects/DVA/Data/UKBiobank/ukb_bgi/ukb_imp_chr", chrom, "_v3.bgen.bgi")
        geno_dat <- bgen.load(filename = bgen_filename, index.filename = bgen_index, rsids = chrom_dat$rsID)
        coeffs <- geno_dat %>%
            pluck("variants") %>%
            left_join(chrom_dat, by = c("rsid" = "rsID")) %>%
            mutate(aligned = case_when(allele1 == effect_allele & allele0 == other_allele ~ list(matrix(c(0,1,2), nrow = 1)),
                                       allele0 == effect_allele & allele1 == other_allele ~ list(matrix(c(2,1,0), nrow = 1)),
                                       TRUE ~ list(NaN)),
                   aligned = map2(effect_weight, aligned, ~.x * .y)) %>%
            pull(aligned) %>%
            reduce(rbind)
        coeffs <- coeffs %>%
            array(dim = c(nrow(coeffs), 1, 3)) %>%
            tensorA::to.tensor() %>%
            tensorA::`names<-.tensor`(c("rsid", "coeff", "probs"))
        geno_dat <- geno_dat %>%
            pluck("data") %>%
            tensorA::to.tensor() %>%
            tensorA::`names<-.tensor`(c("rsid", "samples", "probs")) %>%
            tensorA::mul.tensor("probs", coeffs, "probs", by = "rsid")
        geno_dat <- geno_dat[,1,]
        if(length(dim(geno_dat)) == 2){
            geno_dat <- rowSums(geno_dat, na.rm = TRUE)
        }
        grs_chrom[[as.character(chrom)]] <- geno_dat
    }
    grs_chrom <- reduce(grs_chrom, cbind)
    if(length(dim(grs_chrom)) == 2){
        grs_chrom <- rowSums(grs_chrom, na.rm = TRUE)
    }
    grs_chrom <- data.frame(G = grs_chrom)
    colnames(grs_chrom) <- trait
    write_tsv(grs_chrom, paste0("../data/PRScoresUKB/", trait, ".tsv"))
}       
```

***Note:*** The corresponding IDs for this matrix of genetic scores would be located in:

```{r}
"~/projects/DVA/Data/UKBiobank/ukb_57232/genetic_files/ukb57232_imp_chr1_v3_s487266.sample"
```

